lightning-transaction-sync/src/esplora.rs

```diff
@@ -0,0 +1,378 @@
+	///
+	/// For example, instances of [`ChannelManager`] and [`ChainMonitor`] can be informed about the
+	/// newest on-chain activity related to the items previously registered via the [`Filter`]
+	/// interface.
+	///
+	/// [`Confirm`]: lightning::chain::Confirm
+	/// [`ChainMonitor`]: lightning::chain::chainmonitor::ChainMonitor
+	/// [`ChannelManager`]: lightning::ln::channelmanager::ChannelManager
+	/// [`Filter`]: lightning::chain::Filter
+	#[maybe_async]
+	pub fn sync(&self, confirmables: Vec<&(dyn Confirm + Sync + Send)>) -> Result<(), TxSyncError> {
+		// This lock makes sure we're syncing once at a time.
+		#[cfg(not(feature = "async-interface"))]
+		let mut sync_state = self.sync_state.lock().unwrap();
+		#[cfg(feature = "async-interface")]
+		let mut sync_state = self.sync_state.lock().await;
+
+		log_info!(self.logger, "Starting transaction sync.");
+
+		let mut tip_hash = maybe_await!(self.client.get_tip_hash())?;
+
+		loop {
+			let pending_registrations = self.queue.lock().unwrap().process_queues(&mut sync_state);
+			let tip_is_new = Some(tip_hash) != sync_state.last_sync_hash;
+
+			// We loop until any registered transactions have been processed at least once, or the
+			// tip hasn't been updated during the last iteration.
+			if !sync_state.pending_sync && !pending_registrations && !tip_is_new {
+				// Nothing to do.
+				break;
+			} else {
+				// Update the known tip to the newest one.
+				if tip_is_new {
+					// First check for any unconfirmed transactions and act on it immediately.
+					match maybe_await!(self.get_unconfirmed_transactions(&confirmables)) {
+						Ok(unconfirmed_txs) => {
+							// Double-check the tip hash. If if it changed, a reorg happened since
+							// we started syncing and we need to restart last-minute.
+							let check_tip_hash = maybe_await!(self.client.get_tip_hash())?;
+							if check_tip_hash != tip_hash {
+								tip_hash = check_tip_hash;
+								continue;
+							}
+
+							self.sync_unconfirmed_transactions(&mut sync_state, &confirmables, unconfirmed_txs);
+						},
+						Err(err) => {
+							// (Semi-)permanent failure, retry later.
+							log_error!(self.logger, "Failed during transaction sync, aborting.");
+							sync_state.pending_sync = true;
+							return Err(TxSyncError::from(err));
+						}
+					}
+
+					match maybe_await!(self.sync_best_block_updated(&confirmables, &tip_hash)) {
+						Ok(()) => {}
+						Err(InternalError::Inconsistency) => {
+							// Immediately restart syncing when we encounter any inconsistencies.
+							log_debug!(self.logger, "Encountered inconsistency during transaction sync, restarting.");
+							sync_state.pending_sync = true;
+							continue;
+						}
+						Err(err) => {
+							// (Semi-)permanent failure, retry later.
+							sync_state.pending_sync = true;
+							return Err(TxSyncError::from(err));
+						}
+					}
+				}
+
+				match maybe_await!(self.get_confirmed_transactions(&sync_state)) {
+					Ok(confirmed_txs) => {
```

Comment:
Would it make sense to DRY up the error handling?

Reply:
I was thinking more like a separate method with (some of?) the loop body where the call site has the error handling. Could have a separate internal error for the `continue` case.

```rust
let unconfirmed_txs = maybe_await!(self.get_unconfirmed_transactions(&confirmables))?;
// ...
let check_tip_hash = maybe_await!(self.client.get_tip_hash())?;
if check_tip_hash != tip_hash {
    tip_hash = check_tip_hash;
    return Err(InternalError::Reorg);
}

self.sync_unconfirmed_transactions(&mut sync_state, &confirmables, unconfirmed_txs);
maybe_await!(self.sync_best_block_updated(&confirmables, &tip_hash))?;

let confirmed_txs  = maybe_await!(self.get_confirmed_transactions(&sync_state))?;
let check_tip_hash = maybe_await!(self.client.get_tip_hash())?;
if check_tip_hash != tip_hash {
    tip_hash = check_tip_hash;
    return Err(InternalError::Reorg);
}

self.sync_confirmed_transactions(&mut sync_state, &confirmables, confirmed_txs);
```

Would probably have to return the tip or find some way to update that variable. But otherwise the code is very flat and streamlined in comparison. And the error handling would be at the call site on one `match` instead of three.

Reply:
Actually, `check_tip_hash` could be part of the `InternalError::Reorg` enum.

Reply:
Cool, feel free to squash, IMO.